### PR TITLE
8267833: Improve G1CardSetInlinePtr::add()

### PR DESCRIPTION
**Pull request by walulyai**
Hi all,

Please review this cleanup change to reduce the number of comparisons made when adding entries to an inline pointer cardset container. We allow the find to restart from a previously known index.

Testing: Tier 1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267833](https://bugs.openjdk.java.net/browse/JDK-8267833): Improve G1CardSetInlinePtr::add()

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to e271088aa218475ebd93b52846fcee30d403f792
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5067/head:pull/5067` \
`$ git checkout pull/5067`

Update a local copy of the PR: \
`$ git checkout pull/5067` \
`$ git pull https://git.openjdk.java.net/jdk pull/5067/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5067`

View PR using the GUI difftool: \
`$ git pr show -t 5067`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5067.diff">https://git.openjdk.java.net/jdk/pull/5067.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — APPROVED** [review]
Lgtm.

File: src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp [hunk 1]
```diff
@@ -47,10 +47,14 @@ inline G1CardSetInlinePtr::CardSetPtr G1CardSetInlinePtr::merge(CardSetPtr orig_
 inline G1AddCardResult G1CardSetInlinePtr::add(uint card_idx, uint bits_per_card, uint max_cards_in_inline_ptr) {
   assert(_value_addr != nullptr, "No value address available, cannot add to set.");
 
+  uint cur_idx = 0;
   while (true) {
     uint num_elems = num_cards_in(_value);
+    if (num_elems > 0) {
+      cur_idx = find(card_idx, bits_per_card, cur_idx, num_elems);
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp:54]: I found the naming here a bit confusing, with `num_elems` here but `end_at` in the find function.  Similarly in `contains`.

File: src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp [hunk 2]
```diff
@@ -72,19 +76,29 @@ inline G1AddCardResult G1CardSetInlinePtr::add(uint card_idx, uint bits_per_card
...
     }
     value >>= bits_per_card;
   }
-  return false;
+  return end_at;
+}
+
+inline bool G1CardSetInlinePtr::contains(uint card_idx, uint bits_per_card) {
+  uint num_elems = num_cards_in(_value);
+  if(num_elems == 0) {
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp:97]: Space after `if`

File: src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp [hunk 3]
```diff
@@ -72,19 +76,29 @@ inline G1AddCardResult G1CardSetInlinePtr::add(uint card_idx, uint bits_per_card
...
-      return true;
+      return cur_idx;
     }
     value >>= bits_per_card;
   }
-  return false;
+  return end_at;
+}
+
+inline bool G1CardSetInlinePtr::contains(uint card_idx, uint bits_per_card) {
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp:95]: I think `contains` previously didn't use it's `card_idx` argument.  Is it needed by any callers?  Or was the previous non-use just a performance bug?

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp (hunk 3 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp:95]: D'oh!  Of course it's used.  It's the thing we're looking for.  I just couldn't see it right there in front of me.  Never mind.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp (hunk 1 above)
Comment (walulyai) [reply src/hotspot/share/gc/g1/g1CardSetContainers.inline.hpp:54]: Fixed, opted to use `num_elems` in find function too.

**Review by kimbarrett — APPROVED** [review]

**Comment by walulyai** [comment]
Thanks @kimbarrett and @tschatzl for the reviews!

/integrate

**Pull request closed without merging**